daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala

```diff
@@ -111,6 +111,17 @@ class ContractStateMachine[Nid](mode: ContractKeyUniquenessMode) {
 
     def mode: ContractKeyUniquenessMode = ContractStateMachine.this.mode
 
+    /** Lookup the given key k. Returns
+      * - Some(KeyActive(cid)) if k maps to cid and cid is active.
+      * - Some(KeyInactive) if there is no active contract with the given key.
+      * - None if we know no mapping for that key.
+      */
+    def lookupActiveKey(key: GlobalKey): Option[ContractStateMachine.KeyMapping] =
+      activeState.getLocalActiveKey(key) match {
+        case None => lookupActiveGlobalKeyInput(key)
+        case Some(mapping) => Some(mapping)
+      }
+
     def lookupActiveGlobalKeyInput(key: GlobalKey): Option[ContractStateMachine.KeyMapping] =
       globalKeyInputs.get(key).map {
         case Transaction.KeyActive(cid) if !activeState.consumedBy.contains(cid) =>
```

Comment:
I propose we fix the mess I have done with `Transaction.KeyActive`. 

Either we drop it and use directly `Some` and `None`, either we create a propose algebraic type for it.

Reply:
@andreaslochbihler-da do you have a preference ?

Reply:
I'd prefer a dedicated algebraic type. Because we're stuffing these values into maps and then we have to worry about `Option[Option[ContractId]]`, which is much harder to understand than `Option[Keymapping]`.

Reply:
I think directly using `Some` and `None` is even more confusing. Then you end up with things like `Some(None)` and noone has any clue what’s going on. My preference would be a separate ADT. I’m not keen on including this in this PR though. How about we keep it for the backport and for `main`, I either make a PR as a prerequisite to this or a succeeding this that changes this?

Reply:
Happy with a new ADT, 
Fine with a separate PR as well.

Reply:
Great, I’ll do it as a follow-up to this then. Should get to it today or Monday at the latest. Anything else you’d like me to address before merging?

Reply:
nope.